### PR TITLE
SVDSBTL: process-wide bounded LRU cache for in-memory surfaces

### PR DESCRIPTION
**Pull request by jowr**
### Requirements

* Fill out this template to the extent possible so that this PR can be reviewed in a timely manner.
* Replace the bracketed text below with your own.
* All new code requires tests to ensure against regressions.

### Description of the Change

Every SVDSBTLBackend construction re-read and re-deserialized the .svd.bin.z disk cache (~80ms/surface) even when an identical fluid/source/input-pair/options surface was already resident from a prior AbstractState in the same process. Add SVDSurfaceCache, a thread-safe singleton keyed by the on-disk cache path, that sits in front of ensure_surface_()'s disk load so repeat requests are a map lookup instead of a disk read + msgpack/zlib deserialize.

The cache is bounded by both an entry count and a total estimated byte-size budget (SVDSBTL_SURFACE_CACHE_MAX_ENTRIES / SVDSBTL_SURFACE_CACHE_MAX_SIZE_MB config keys, default 16 / 512 MB), evicting least-recently-used entries first when either limit is hit -- surface size varies widely with the NT/NR/rank build options, so an entry-count cap alone can't bound memory. surfaces_ now holds shared_ptr<const SVDSurface> instead of unique_ptr so a cache hit can be shared read-only across backend instances.

### Benefits

* Eliminates the redundant ~80 ms disk read + msgpack/zlib deserialize on every `SVDSBTLBackend` construction for a `(fluid, source, input_pair, options)` tuple that's already resident in the process -- a second `AbstractState::factory("SVDSBTL&...", ...)` call for the same surface becomes a map lookup.
* Directly benefits any call pattern that rebuilds the backend per call against a small, repeated set of fluids -- most notably `PropsSI`, whose docs/conformance-script use case was the original motivation for the `ALLOW_SVDSBTL_IN_PROPSSI` gate.
* Memory growth is bounded and configurable (`SVDSBTL_SURFACE_CACHE_MAX_ENTRIES` / `SVDSBTL_SURFACE_CACHE_MAX_SIZE_MB`, both with sane defaults), with LRU eviction, so the cache is safe to leave enabled by default rather than requiring opt-in.

### Possible Drawbacks

* Adds a process-wide mutable singleton (mutex-protected) where none existed before. Sharing relies on `SVDSurface` staying immutable after `seal()` -- true today (verified: `add_region`/`add_region_property_svd`/`seal` are the only mutators and are only ever called before a surface is published to the cache), but any future change that mutates a sealed `SVDSurface` in place (e.g. via `const_cast`) would silently become unsafe across threads/instances.
* `estimate_surface_bytes()` sums only the `SVDDecomposition` vectors (the dominant cost at production grid sizes); it does not account for `RegionAtlas` / `SVDEvaluator` bookkeeping overhead, so `SVDSBTL_SURFACE_CACHE_MAX_SIZE_MB` is an approximate, not exact, resident-memory bound.
* Eviction only runs inside `put()`. Lowering the config caps at runtime does not shrink an already-oversized cache until the next surface is loaded (a `get()`-only workload after a config change keeps the old, larger footprint). Documented as a known trade-off in the code; matches the existing `SVDSBTL_SAMPLING_THREADS` live-config-read convention elsewhere in this backend.
* Default budgets (16 entries / 512 MB) are reasoned estimates, not tuned against a specific deployment's memory budget; very memory-constrained environments may want to lower them explicitly.

### Verification Process

* Checked the project's `bd` issue tracker first for an existing/duplicate effort before implementing -- none found (closest is `CoolProp-eal`, lazy per-pair surface loading, which is complementary, not overlapping).
* Built the full project locally (`cmake -B build_catch -S . -DCOOLPROP_CATCH_MODULE=ON -DBUILD_TESTING=ON -DCMAKE_BUILD_TYPE=Release` + `cmake --build build_catch --target CatchTestRunner`) and ran the `[SBTL]` umbrella tag per this repo's test-filter discipline (touches `src/Backends/SVDSBTL/`).
* **New functionality:** added 4 new fast, disk/HEOS-free unit tests directly against `SVDSurfaceCache` in `CoolProp-Tests-SBTLAdapter.cpp`:
  - `estimate_surface_bytes` sums decomposition vectors across regions/properties (verifies the byte-accounting math against a hand-fabricated surface).
  - `SVDSurfaceCache get/put round-trip and miss`.
  - `SVDSurfaceCache evicts least-recently-used entry once max entry count is exceeded` (asserts the *specific* victim, not just a count).
  - `SVDSurfaceCache evicts oldest entry once the byte-size budget is exceeded` (sizes the config cap to admit exactly one ~3 MB fabricated surface, not two, and asserts the specific victim).
* **Changed functionality / regressions:** ran the full `[SBTL]` suite four times across the change's lifecycle (initial implementation, after adding the new cache tests, after a clang-format pass, and after the doc/comment sync) -- consistently **36 test cases, 5474/5474 assertions passed, 1 expected skip** (REFPROP not installed in the build sandbox). This includes the two pre-existing tests whose names reference a "surface cache" (`SVDSBTL DmassT surface caches + reloads from disk (CoolProp-i7j)` and `SVDSBTL prebuild shares the surface cache with a plain instance`) -- both are about the pre-existing *on-disk* cache, unaffected by this change, confirmed still passing.
* Ran `clang-format-18` (the version pinned in `.pre-commit-config.yaml`) over every changed/new file; zero diff.
* Per this repo's mandatory pre-PR gate, ran an adversarial code-review subagent against the full diff, specifically probing thread-safety (mutex coverage, iterator invalidation in the LRU list), LRU/eviction correctness (off-by-one, byte double-counting on replace, the `MAX_ENTRIES=0` "disables the cache" edge case), the safety of sharing `shared_ptr<const SVDSurface>` across backend instances, and whether the new unit tests would actually fail if the eviction logic were broken. No blocking findings.

### Applicable Issues

None open on GitHub. Cross-checked against this project's `bd` issue tracker beforehand; no existing issue duplicates this work. (Related, not overlapping: `CoolProp-eal` tracks making per-pair surface *loading* lazy, independent of this in-memory *reuse* cache.)

---
_Generated by [Claude Code](https://claude.ai/code/session_01LumnxFsKp5mZmkE9n6ewCP)_

## Summary by CodeRabbit

* **New Features**
  * Added a process-wide, thread-safe in-memory surface cache for SVDSBTL to reduce repeated disk loads and deserialization within the same process.
  * Added configuration keys to bound the cache by maximum entries and total resident size.
* **Bug Fixes**
  * Improved repeated SVDSBTL performance for identical fluid/options by reusing cached surfaces across backend instances.
* **Documentation**
  * Updated low-level guidance to clarify required instance allocation behavior and SVDSBTL/`PropsSI` performance with the new cache.
* **Tests**
  * Added unit tests covering cache behavior (miss/hit), shared instance reuse, and LRU/size-based eviction.

### CODE REVIEW

File: src/Backends/SVDSBTL/SVDSBTLBackend.cpp [hunk 1]
```diff
@@ -996,11 +1002,18 @@
...
+    // the msgpack/zlib deserialize that load_from_file() would
+    // otherwise pay on every backend construction (the dominant cost
+    // for PropsSI-style call patterns that rebuild the backend per
+    // call).
+    std::shared_ptr<const cp_sbtl::SVDSurface> surface = cp_sbtl::SVDSurfaceCache::instance().get(path);
+
+    if (!surface && std::filesystem::exists(path)) {
         try {
-            surface = std::make_unique<cp_sbtl::SVDSurface>(cp_sbtl::SVDSurfaceSerializer::load_from_file(path));
+            surface = std::make_shared<cp_sbtl::SVDSurface>(cp_sbtl::SVDSurfaceSerializer::load_from_file(path));
```
[the remark this answers is not in the document]
Comment (jowr) [reply src/Backends/SVDSBTL/SVDSBTLBackend.cpp:1021]: Looked into this — I believe it's a pre-existing pattern misattributed as "new" by the PR-diff view, not a vulnerability introduced by this change.

`load_from_file(path)` at this line is unchanged by this PR (the only edit here was adding an in-memory cache lookup a few lines above it). `path` is built by `SVDSurfaceSerializer::default_cache_path()` / `default_cache_dir()` in `src/SBTL/SVDSurfaceSerializer.cpp` — a file this PR doesn't touch. The taint source CodeQL is tracing is almost certainly `$HOME` (via `get_home_dir()`) or the `ALTERNATIVE_SVDTABLES_DIRECTORY` config key used to build the cache directory root; that logic, and this exact call site, already exist unchanged on `master`. This alert is showing up as "new in this PR" only because the surrounding lines in `ensure_surface_()` shifted in the diff.

`default_cache_path()` already validates `fluid_name`/`source_backend`/`opthash` against path-separator and `..` characters (defense-in-depth against a caller passing an attacker-controlled fluid name). The directory root itself (`$HOME` / `ALTERNATIVE_SVDTABLES_DIRECTORY`) isn't separately sanitized for traversal segments, but in CoolProp's threat model (a local library, not a network-facing service) both are set by whoever is already running the process — not an external input crossing a trust boundary — so I don't think this is exploitable here.

Leaving this for a maintainer to triage/dismiss on the code-scanning page directly, since a real hardening fix (if wanted) would land in `SVDSurfaceSerializer.cpp`'s `default_cache_dir()`, which is out of scope for this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by ibell** [comment]
Looks pretty good to me, nothing that gives me concern!

**Comment by jowr** [comment]
Both CodeRabbit nitpicks are addressed in 001d489, and master has been merged in (af0eb2c) so the branch is no longer 30 commits behind.

**1. Exception-safe test cleanup** — valid, and the more consequential of the two. Both eviction tests mutated the global cache-limit config keys plus the process-wide singleton and restored them only at the end of the test body; since `REQUIRE` throws, a failing assertion skipped restoration and leaked the mutated limits into every subsequent test in the binary, turning one real failure into a cascade. The byte-budget case was the clearer instance — it sets `MAX_ENTRIES` before a `REQUIRE`. Restoration now lives in a `CacheConfigGuard` destructor (also clears the cache), so it runs on any exit path.

**2. Redundant `put()` on a cache hit** — also valid. `ensure_surface_()` published unconditionally, so a hit paid a second lock, a byte re-estimate, an erase/reinsert of the identical entry, and an eviction sweep, all redundant given `get()` already promotes to MRU. Now gated on a `from_memory_cache` flag so only freshly loaded/rebuilt surfaces are published.

**Merge check:** no textual conflicts. One semantic risk worth noting for the record — master's #3252 regenerated `wrappers/Python/_nanobind/CoolProp.pyi`, the same file the earlier stub fix here touched. Master's addition is in the `AbstractState` method block and this branch's is in the `configuration_keys` enum, both purely additive in disjoint regions, so the merged stub should still match a fresh stubgen run; `configuration_keys.h` itself was untouched on master, so the enum values (38/39) remain correct. The `nanobind wheels` gate re-running on this push will confirm.

Re-verified on the merged tree: full build clean, `[SBTL]` suite 36 cases / 5474 assertions passing (1 expected skip — no REFPROP in the build sandbox), `clang-format-18` zero diff.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: include/CoolProp/Backends/SVDSBTL/SVDSBTLBackend.h [hunk 2]
```diff
@@ -342,8 +356,10 @@ class SVDSBTLBackend : public AbstractState
     void ensure_dome_s_endpoints_(PointEvaluation& pt);
     void ensure_dome_h_endpoints_(PointEvaluation& pt);
 
-    // Load <fluid>.<pair>.svd.bin.z from the default cache; if absent,
-    // build via the matching preset and save it.  Inserts into surfaces_.
+    // Resolve the surface for `pair`: first the process-wide
+    // SVDSurfaceCache, then <fluid>.<pair>.svd.bin.z on disk; if
+    // neither has it, build via the matching preset and populate
+    // both.  Inserts into surfaces_.
```
[the remark this answers is not in the document]
Comment (jowr) [reply include/CoolProp/Backends/SVDSBTL/SVDSBTLBackend.h]: Good catch — fixed in 417a8f0.

Agreed this is worth more than a tidiness fix: `SVDSurfaceCache` keys on the same full path string, so the abbreviated `<fluid>.<pair>.svd.bin.z` really could read as though `SVDSBTL&HEOS` and `SVDSBTL&REFPROP` (or two different `grid`/`rank` option sets) share one entry — which would contradict the correctness argument for the whole feature.

I wrote out the concrete format rather than the suggested "full source/options-specific cache path", since being explicit seemed more useful to a future reader:

```
// Resolve the surface for `pair`: first the process-wide
// SVDSurfaceCache, then the on-disk cache at
// <fluid>.<source>.<input_pair>.<opthash>.svd.bin.z (see
// SVDSurfaceSerializer::default_cache_path); if neither has it,
// build via the matching preset and populate both.  Inserts into
// surfaces_.
//
// The in-memory cache is keyed by that same full path string, so
// differing source backends or option sets never share an entry.
```

Comment-only change; rebuilt to confirm it still compiles, `clang-format-18` clean.

---
_Generated by [Claude Code](https://claude.ai/code)_
